discovery/wh_discovery.go

```diff
@@ -0,0 +1,162 @@
+package discovery
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/livepeer/go-livepeer/core"
+	"github.com/livepeer/go-livepeer/net"
+
+	"github.com/golang/glog"
+)
+
+var whRefreshInterval = 1 * time.Minute
+
+type webhookResponse struct {
+	Address string
+}
+
+type webhookPool struct {
+	node         *core.LivepeerNode
+	pool         *orchestratorPool
+	callback     *url.URL
+	responseHash ethcommon.Hash
+	lastRequest  time.Time
+	mu           *sync.RWMutex
+}
+
+func NewWebhookPool(node *core.LivepeerNode, callback *url.URL) *webhookPool {
+	body, err := getURLsfromWebhook(callback)
+	if err != nil {
+		glog.Error(fmt.Errorf("Error retriving urls from webhook: %v", err))
+		return nil
+	}
+
+	addrs, err := deserializeWebhookJSON(body)
+	if err != nil {
+		glog.Error(fmt.Errorf("Error caching urls from webhook: %v", err))
+		return nil
+	}
+
+	pool := NewOrchestratorPool(node, addrs)
+
+	return &webhookPool{
+		pool:         pool,
+		node:         node,
+		callback:     callback,
+		lastRequest:  time.Now(),
+		responseHash: ethcommon.BytesToHash(crypto.Keccak256(body)),
+		mu:           &sync.RWMutex{},
+	}
+}
+
+func (w *webhookPool) getURLs() ([]*url.URL, error) {
```

Comment:
This function doesn't really assure mutual exclusivity of the webhook call - multiple calls could be occurring at once, with "last response wins" - but it's probably OK for now.

Reply:
oh yes, that's a part I didn't add. If it's okay for now, I'd rather leave it as is and make note of that as an optimization for the future